File: engine.py
"""Движок: периодический «тик», который решает, что и когда отправить.

Запускается раз в минуту из job_queue. Идемпотентен — состояние (что уже
отправлено, когда слали отчёты) держится в базе, поэтому повторный тик в ту же
минуту ничего не задваивает.

Реализует правила слоя 2 поверх хранилища:
  * кадентность и next_fire_at по треку;
  * тишина в воскресенье (кроме отчётов);
  * приоритет: не больше одного напоминания за тик (6min раньше CSCA раньше ЕГЭ),
    чтобы не приходило «оба разом»;
  * правило пропусков: 2 пропуска подряд -> вместо напоминания вопрос «что случилось».
"""

from telegram.constants import ParseMode

import config
import reminders
import scheduling
import stats
from db import from_iso, utcnow


async def run_tick(bot, store):
    for user in store.all_users():
        try:
            await _process_user(bot, store, user)
        except Exception as e:  # один пользователь не должен ронять цикл
            print(f"[tick] user {user['user_id']} error: {e}")


async def _resume_if_pause_over(store, user):
    if user["active"]:
        return True
    pu = user["paused_until"]
    if pu and utcnow() >= from_iso(pu):
        store.set_active(user["user_id"], 1, None)
        return True
    return False


async def _process_user(bot, store, user):
    uid = user["user_id"]
    tz = user["tz"]
    now_local = scheduling.local_now(tz)

    active = await _resume_if_pause_over(store, user)

    # Отчёты идут даже в воскресенье и на паузе? На паузе — молчим совсем.
    if not active:
        return

    await _maybe_reports(bot, store, user, now_local)

    # Правило воскресенья: напоминаний нет.
    if scheduling.is_quiet_day(now_local):
        return

    # Собираем сработавшие треки в порядке приоритета.
    due_tracks = []
    for track in store.get_tracks(uid):
        if not track["enabled"]:
            continue
        nf = from_iso(track["next_fire_at"])
        if nf is None:
            first = scheduling.compute_first_fire(track, tz)
            store.set_next_fire(uid, track["track_id"], first)
            continue
        if utcnow() >= nf:
            due_tracks.append(track)

    if not due_tracks:
        return

    ordered = sorted(
        due_tracks,
        key=lambda t: config.TRACK_ORDER.index(t["track_id"])
        if t["track_id"] in config.TRACK_ORDER else 99,
    )

    # Не больше одного за тик — «не оба разом».
    track = ordered[0]
    await _send_reminder(bot, store, user, track, now_local)


async def _send_reminder(bot, store, user, track, now_local):
    uid = user["user_id"]
    tid = track["track_id"]
    sched_date = now_local.date().isoformat()

    # Уже слали сегодня по этому треку? Тогда просто сдвигаем next_fire вперёд.
    existing = store.pending_today(uid, tid, sched_date)
    if existing:
        nxt = scheduling.compute_next_fire(track, user["tz"],
                                           scheduling.local_now(user["tz"]))
        store.set_next_fire(uid, tid, nxt)
        return

    idx = track["content_index"]

    # Правило пропусков: 2 подряд 'skipped' -> вопрос вместо напоминания.
    last = store.last_terminal(uid, tid, limit=config.SKIP_STREAK_THRESHOLD)
    skip_streak = (
        len(last) >= config.SKIP_STREAK_THRESHOLD
        and all(r["status"] == "skipped" for r in last)
    )

    if skip_streak:
        hid = store.add_history(uid, tid, sched_date, utcnow(),
                                "[вопрос: что случилось]", "", "sent")
        text, kb = reminders.build_skip_question(tid, hid)
        await bot.send_message(user["chat_id"], text, parse_mode=ParseMode.MARKDOWN,
                               reply_markup=kb)
    else:
        text, link, title, minutes = reminders.build_reminder(tid, idx)
        hid = store.add_history(uid, tid, sched_date, utcnow(), title, link, "sent")
        kb = reminders.action_keyboard(hid, link)
        await bot.send_message(user["chat_id"], text, parse_mode=ParseMode.MARKDOWN,
                               reply_markup=kb, disable_web_page_preview=False)
        store.bump_content_index(uid, tid)

    # Двигаем следующее срабатывание по кадентности.
    nxt = scheduling.compute_next_fire(track, user["tz"], now_local)
    store.set_next_fire(uid, tid, nxt)


async def _maybe_reports(bot, store, user, now_local):
    uid = user["user_id"]
    tz = user["tz"]

    # Недельный: воскресенье, вечер, один раз в неделю.
    if (now_local.weekday() == config.WEEKLY_REPORT_WEEKDAY
            and now_local.hour >= config.WEEKLY_REPORT_HOUR):
        wk = now_local.isocalendar()
        tag = f"{wk[0]}-W{wk[1]}"
        if store.get_meta(uid, "last_weekly") != tag:
            text = stats.weekly_report(store, uid, tz)
            await bot.send_message(user["chat_id"], text,
                                   parse_mode=ParseMode.MARKDOWN)
            store.set_meta(uid, "last_weekly", tag)

    # Месячный: 1-е число, утро, один раз в месяц.
    if (now_local.day == config.MONTHLY_REPORT_DAY
            and now_local.hour >= config.MONTHLY_REPORT_HOUR):
        tag = f"{now_local.year}-{now_local.month:02d}"
        if store.get_meta(uid, "last_monthly") != tag:
            text, suggestion = stats.monthly_report(store, uid, tz)
            kb = None
            if suggestion:
                from telegram import InlineKeyboardButton, InlineKeyboardMarkup
                sid, new_n = suggestion
                kb = InlineKeyboardMarkup([[
                    InlineKeyboardButton(
                        f"✅ Да, реже (раз в {new_n} дн.)",
                        callback_data=f"adapt:{sid}:{new_n}"),
                    InlineKeyboardButton("Оставить", callback_data="adapt:none:0"),
                ]])
            await bot.send_message(user["chat_id"], text,
                                   parse_mode=ParseMode.MARKDOWN, reply_markup=kb)
            store.set_meta(uid, "last_monthly", tag)

File: worker/src/engine.js
// Движок: scheduled() дёргает runTick раз в минуту. Идемпотентен —
// состояние в D1, повторный тик ничего не задваивает.

import {
  TRACK_ORDER, QUIET_WEEKDAY, SKIP_STREAK_THRESHOLD,
  WEEKLY_REPORT_WEEKDAY, WEEKLY_REPORT_HOUR,
  MONTHLY_REPORT_DAY, MONTHLY_REPORT_HOUR, TRACK_TITLES,
} from "./config.js";
import * as db from "./db.js";
import * as sched from "./scheduling.js";
import * as rem from "./reminders.js";
import * as stats from "./stats.js";
import { sendMessage } from "./telegram.js";

export async function runTick(env) {
  const users = await db.allUsers(env);
  for (const user of users) {
    try {
      await processUser(env, user);
    } catch (e) {
      console.log(`tick user ${user.user_id} error: ${e && e.stack || e}`);
    }
  }
}

async function resumeIfPauseOver(env, user) {
  if (user.active) return true;
  if (user.paused_until && Date.now() >= user.paused_until) {
    await db.setActive(env, user.user_id, 1, null);
    return true;
  }
  return false;
}

async function processUser(env, user) {
  const uid = user.user_id;
  const tz = user.tz;
  const now = Date.now();

  const active = await resumeIfPauseOver(env, user);
  if (!active) return;

  await maybeReports(env, user, now);

  if (sched.isQuietDay(now, tz)) return; // воскресенье — тишина

  const tracks = await db.getTracks(env, uid);
  const due = [];
  for (const track of tracks) {
    if (!track.enabled) continue;
    if (track.next_fire_at == null) {
      await db.setNextFire(env, uid, track.track_id, sched.computeFirstFire(track, tz, now));
      continue;
    }
    if (now >= track.next_fire_at) due.push(track);
  }
  if (!due.length) return;

  due.sort((a, b) => {
    const ia = TRACK_ORDER.indexOf(a.track_id), ib = TRACK_ORDER.indexOf(b.track_id);
    return (ia < 0 ? 99 : ia) - (ib < 0 ? 99 : ib);
  });

  // Не больше одного за тик — «не оба разом».
  await sendReminder(env, user, due[0], now);
}

async function sendReminder(env, user, track, now) {
  const uid = user.user_id;
  const tid = track.track_id;
  const tz = user.tz;
  const dateStr = sched.localInfo(now, tz).dateStr;

  const existing = await db.pendingToday(env, uid, tid, dateStr);
  if (existing) {
    await db.setNextFire(env, uid, tid, sched.computeNextFire(track, tz, now));
    return;
  }

  const last = await db.lastTerminal(env, uid, tid, SKIP_STREAK_THRESHOLD);
  const skipStreak = last.length >= SKIP_STREAK_THRESHOLD
    && last.every((r) => r.status === "skipped");

  if (skipStreak) {
    const hid = await db.addHistory(env, uid, tid, dateStr, "[вопрос: что случилось]", "", "sent");
    const { text, kb } = rem.buildSkipQuestion(tid, hid);
    await sendMessage(env, user.chat_id, text, { replyMarkup: kb });
  } else {
    const { text, link, title } = rem.buildReminder(tid, track.content_index);
    const hid = await db.addHistory(env, uid, tid, dateStr, title, link, "sent");
    await sendMessage(env, user.chat_id, text, { replyMarkup: rem.actionKeyboard(hid, link) });
    await db.bumpContentIndex(env, uid, tid);
  }

  await db.setNextFire(env, uid, tid, sched.computeNextFire(track, tz, now));
}

async function maybeReports(env, user, now) {
  const uid = user.user_id;
  const tz = user.tz;
  const info = sched.localInfo(now, tz);

  if (info.weekday === WEEKLY_REPORT_WEEKDAY && info.hh >= WEEKLY_REPORT_HOUR) {
    const tag = isoWeekTag(info);
    if ((await db.getMeta(env, uid, "last_weekly")) !== tag) {
      const text = await stats.weeklyReport(env, uid, tz);
      await sendMessage(env, user.chat_id, text);
      await db.setMeta(env, uid, "last_weekly", tag);
    }
  }

  if (info.d === MONTHLY_REPORT_DAY && info.hh >= MONTHLY_REPORT_HOUR) {
    const tag = `${info.y}-${String(info.m).padStart(2, "0")}`;
    if ((await db.getMeta(env, uid, "last_monthly")) !== tag) {
      const { text, suggestion } = await stats.monthlyReport(env, uid, tz);
      let replyMarkup;
      if (suggestion) {
        replyMarkup = {
          inline_keyboard: [[
            { text: `✅ Да, реже (раз в ${suggestion.newN} дн.)`, callback_data: `adapt:${suggestion.tid}:${suggestion.newN}` },
            { text: "Оставить", callback_data: "adapt:none:0" },
          ]],
        };
      }
      await sendMessage(env, user.chat_id, text, { replyMarkup });
      await db.setMeta(env, uid, "last_monthly", tag);
    }
  }
}

// Тег ISO-недели (год + номер недели) для дедупликации недельного отчёта.
function isoWeekTag(info) {
  const d = new Date(Date.UTC(info.y, info.m - 1, info.d));
  const day = (d.getUTCDay() + 6) % 7;
  d.setUTCDate(d.getUTCDate() - day + 3);
  const firstThu = new Date(Date.UTC(d.getUTCFullYear(), 0, 4));
  const week = 1 + Math.round(
    ((d - firstThu) / 86400000 - 3 + ((firstThu.getUTCDay() + 6) % 7)) / 7
  );
  return `${d.getUTCFullYear()}-W${week}`;
}
